doc/developers/contributing.rst

```diff
@@ -207,6 +207,104 @@ and Cython optimizations.
+
+If you changed the same part of the same file differently in the two branches you’re merging
+together, Git won’t be able to merge them cleanly. This is called a merge conflict. To resolve a
+merge conflict you can either:
+	
+	* Open the files manually and resolve them as git adds standard resolution markers to the
+	  files that have conflicts.
+
+	* Use a graphical tool by running ``git mergetool``. This tool will walk you through all the
+	  conflicts.
+
+.. note::
+	Detailed guide on solving merge conflicts can be found on `git-scm.com
+	<http://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging#Basic-Merge-Conflicts>`_
+
+You should generally use ``rebase`` over ``merge`` to integrate changes from one branch into
+another branch. Merging creates a "merge commit" every time you incorporate upstream changes.
+This pollutes the feature branch's history and makes it hard for other developers to understand
+the history of the project. Whereas ``rebase`` effectively rewrites the project history by creating
+brand new commits for each commit in the original branch. This creates a much cleaner project history.
+To rebase:
+
+	1. Switch to feature branch::
+
+		$ git checkout my-feature-branch
+
+	2. Make a backup just in case::
+
+		$ git branch my-feature-branch-bak my-feature-branch
+
+	3. Rebase on master::
+
+		$ git rebase master
+
+This can create merge conflicts and you might need to resolve them. If successfully rebased
+you can delete the backup branch by::
+
+	$ git branch -D my-feature-branch-bak
+
+Squashing commits
+^^^^^^^^^^^^^^^^^
+
+To keep the history clean and to help backtrack regressions to a single commit we
+prefer less (but meaningful) number of commits.
+Please try to avoid merging master into a feature / bugfix branch as it makes it hard
+to rebase and squash intermediate commits later. Ideally it would be great to start
+branch again off of current master, ``git cherry-pick`` the commits that are needed
+from ``my-feature-branch``, and then ``push -f`` the result of that operation to
+``my-feature-branch``. Squashing commits is done as follows:
+
+	1. Rename your current feature branch in case something goes wrong::
+
+		$ git branch -m my-feature-branch my-feature-branch-bak
+
+	2. Update local master::
+
+		$ git checkout master
+		$ git pull upstream
+
+	3. Delete the ``my-feature-branch``. The code will still be under the branch
+	   ``my-feature-branch-bak`` in case you need it::
+
+		$ git branch -D my-feature-branch
+
+	4. Create a new feature branch (could be named ``my-feature-branch``) and cherry pick
+	   meaningful commits::
+
+		$ git checkout -b new-feature-branch
+		$ git cherry pick *commit-1 id*
+		$ git cherry pick *commit-2 id*
+		.
+		.
+		.
+
+	5. Force push into your PR branch::
+
+		$ git push origin new-feature-branch -f
+
+.. note::
+	Detailed guides on **interactive rebasing** can be found:
+
+	* In the *Scipy Development Workflow* in the `rebasing
```

Comment:
I actually just added the squashing part of that tutorial so though of keeping this link here. I can copy the rest of it too if you want.